third/nmh/man/ali.man
.\"
.\" %nmhwarning%
.\" $Id: ali.man,v 1.1.1.1 1999-02-07 18:14:19 danw Exp $
.\"
.\" include the -mh macro file
.so %etcdir%/tmac.h
.\"
.TH ALI %manext1% MH.6.8 [%nmhversion%]
.SH NAME
ali \- list mail aliases
.SH SYNOPSIS
.in +.5i
.ti -.5i
ali
\%[\-alias\ aliasfile]
\%[\-list] \%[\-nolist]
\%[\-normalize]
.br
\%[\-nonormalize]
\%[\-user] \%[\-nouser]
\%[aliases\ ...]
.br
\%[\-version]
\%[\-help] 
.in -.5i
.SH DESCRIPTION
\fIAli\fR searches the named mail alias files for each of the given
\fIaliases\fR.  It creates a list of addresses for those \fIaliases\fR,
and writes that list on standard output.  If no arguments are given,
\fIali\fR outputs all alias entries.

By default, when an aliases expands to multiple addresses, the addresses
are separated by commas and printed on as few lines as possible.  If the
`\-list' option is specified, then when an address expands to multiple
addresses, each address will appear on a separate line.

The switch `\-user' directs \fIali\fR to perform its processing in
an inverted fashion: instead of listing the addresses that each given
alias expands to, \fIali\fR will list the aliases that expand to each
given address.  If the `\-normalize' switch is given, \fIali\fR will
try to track down the official hostname of the address.

The files specified by the profile entry \*(lqAliasfile:\*(rq and any
additional alias files given by the `\-alias aliasfile' switch will be
read.  Each \fIalias\fR is processed as described in \fImh\-alias\fR\0(5).
.Fi
^$HOME/\&.mh\(ruprofile~^The user profile
^/etc/passwd~^List of users
^/etc/group~^List of groups
.Pr
^Path:~^To determine the user's nmh directory
.Ps
^Aliasfile:~^For a default alias file
.Sa
mh\-alias(5)
.De
`\-alias %etcdir%/MailAliases'
.Ds
`\-nolist'
.Ds
`\-nonormalize'
.Ds
`\-nouser'
.Co
None
.Bu
The `\-user' option with `\-nonormalize' is not entirely accurate, as it
does not replace local nicknames for hosts with their official site names.
.En


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


third/nmh/uip/ali.c

/*
 * ali.c -- list nmh mail aliases
 *
 * $Id: ali.c,v 1.1.1.1 1999-02-07 18:14:12 danw Exp $
 */

#include <h/mh.h>
#include <h/addrsbr.h>
#include <h/aliasbr.h>

/*
 * maximum number of names
 */
#define	NVEC 50

static struct swit switches[] = {
#define	ALIASW                0
    { "alias aliasfile", 0 },
#define	NALIASW               1
    { "noalias", -7 },
#define	LISTSW                2
    { "list", 0 },
#define	NLISTSW               3
    { "nolist", 0 },
#define	NORMSW                4
    { "normalize", 0 },
#define	NNORMSW               5
    { "nonormalize", 0 },
#define	USERSW                6
    { "user", 0 },
#define	NUSERSW               7
    { "nouser", 0 },
#define VERSIONSW             8
    { "version", 0 },
#define	HELPSW                9
    { "help", 4 },
    { NULL, 0 }
};

static int pos = 1;

extern struct aka *akahead;

/*
 * prototypes
 */
void print_aka (char *, int, int);
void print_usr (char *, int, int);


int
main (int argc, char **argv)
{
    int i, vecp = 0, inverted = 0, list = 0;
    int noalias = 0, normalize = AD_NHST;
    char *cp, **ap, **argp, buf[BUFSIZ];
    char *vec[NVEC], **arguments;
    struct aka *ak;

#ifdef LOCALE
    setlocale(LC_ALL, "");
#endif
    invo_name = r1bindex (argv[0], '/');

    /* read user profile/context */
    context_read();

    mts_init (invo_name);
    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW: 
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW: 
		    adios (NULL, "-%s unknown", cp);

		case HELPSW: 
		    snprintf (buf, sizeof(buf), "%s [switches] aliases ...",
			invo_name);
		    print_help (buf, switches, 1);
		    done (1);
		case VERSIONSW:
		    print_version (invo_name);
		    done (1);

		case ALIASW: 
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    if ((i = alias (cp)) != AK_OK)
			adios (NULL, "aliasing error in %s - %s", cp, akerror (i));
		    continue;
		case NALIASW: 
		    noalias++;
		    continue;

		case LISTSW: 
		    list++;
		    continue;
		case NLISTSW: 
		    list = 0;
		    continue;

		case NORMSW: 
		    normalize = AD_HOST;
		    continue;
		case NNORMSW: 
		    normalize = AD_NHST;
		    continue;

		case USERSW: 
		    inverted++;
		    continue;
		case NUSERSW: 
		    inverted = 0;
		    continue;
	    }
	}
	vec[vecp++] = cp;
    }

    if (!noalias) {
	/* allow Aliasfile: profile entry */
	if ((cp = context_find ("Aliasfile"))) {
	    char *dp = NULL;

	    for (ap = brkstring(dp = getcpy(cp), " ", "\n"); ap && *ap; ap++)
		if ((i = alias (*ap)) != AK_OK)
		    adios (NULL, "aliasing error in %s - %s", *ap, akerror (i));
	    if (dp)
		free(dp);
	}
	alias (AliasFile);
    }

    /*
     * If -user is specified
     */
    if (inverted) {
	if (vecp == 0)
	    adios (NULL, "usage: %s -user addresses ...  (you forgot the addresses)",
		   invo_name);

	for (i = 0; i < vecp; i++)
	    print_usr (vec[i], list, normalize);

	done (0);
    }

    if (vecp) {
	/* print specified aliases */
	for (i = 0; i < vecp; i++)
	    print_aka (akvalue (vec[i]), list, 0);
    } else {
	/* print them all */
	for (ak = akahead; ak; ak = ak->ak_next) {
	    printf ("%s: ", ak->ak_name);
	    pos += strlen (ak->ak_name) + 1;
	    print_aka (akresult (ak), list, pos);
	}
    }

    done (0);
}

void
print_aka (char *p, int list, int margin)
{
    char c;

    if (p == NULL) {
	printf ("<empty>\n");
	return;
    }

    while ((c = *p++)) {
	switch (c) {
	    case ',': 
		if (*p) {
		    if (list)
			printf ("\n%*s", margin, "");
		    else {
			if (pos >= 68) {
			    printf (",\n ");
			    pos = 2;
			} else {
			    printf (", ");
			    pos += 2;
			}
		    }
		}

	    case 0: 
		break;

	    default: 
		pos++;
		putchar (c);
	}
    }

    putchar ('\n');
    pos = 1;
}

void
print_usr (char *s, int list, int norm)
{
    register char *cp, *pp, *vp;
    register struct aka *ak;
    register struct mailname *mp, *np;

    if ((pp = getname (s)) == NULL)
	adios (NULL, "no address in \"%s\"", s);
    if ((mp = getm (pp, NULL, 0, norm, NULL)) == NULL)
	adios (NULL, "bad address \"%s\"", s);
    while (getname (""))
	continue;

    vp = NULL;
    for (ak = akahead; ak; ak = ak->ak_next) {
	pp = akresult (ak);
	while ((cp = getname (pp))) {
	    if ((np = getm (cp, NULL, 0, norm, NULL)) == NULL)
		continue;
	    if (!strcasecmp (mp->m_host, np->m_host)
		    && !strcasecmp (mp->m_mbox, np->m_mbox)) {
		vp = vp ? add (ak->ak_name, add (",", vp))
		    : getcpy (ak->ak_name);
		mnfree (np);
		while (getname (""))
		    continue;
		break;
	    }
	    mnfree (np);
	}
    }
    mnfree (mp);

#if 0
    printf ("%s: ", s);
    print_aka (vp ? vp : s, list, pos += strlen (s) + 1);
#else
    print_aka (vp ? vp : s, list, 0);
#endif

    if (vp)
	free (vp);
}
